File: src/main/java/org/distril/beengine/inventory/Inventory.java
package org.distril.beengine.inventory;

import com.nukkitx.protocol.bedrock.packet.ContainerClosePacket;
import com.nukkitx.protocol.bedrock.packet.InventoryContentPacket;
import com.nukkitx.protocol.bedrock.packet.InventorySlotPacket;
import lombok.AccessLevel;
import lombok.Getter;
import org.distril.beengine.material.Material;
import org.distril.beengine.material.item.Item;
import org.distril.beengine.player.Player;
import org.distril.beengine.util.ItemUtils;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.atomic.AtomicInteger;

@Getter
public abstract class Inventory {

	private static final AtomicInteger NEXT_ID = new AtomicInteger(0);

	private final Set<Player> viewers = new HashSet<>();

	private final InventoryHolder holder;
	private final InventoryType type;

	private final Item[] items;

	@Getter(AccessLevel.PROTECTED)
	private final int id;

	public Inventory(InventoryHolder holder, InventoryType type) {
		this(holder, type, null);
	}

	public Inventory(InventoryHolder holder, InventoryType type, Integer overrideId) {
		this.holder = holder;
		this.type = type;
		this.items = new Item[type.getSize()];

		if (overrideId != null) {
			this.id = overrideId;
		} else {
			this.id = NEXT_ID.incrementAndGet();
		}
	}

	public boolean setItem(int slot, Item item) {
		return this.setItem(slot, item, true);
	}

	public boolean setItem(int slot, Item item, boolean send) {
		if (slot < 0 || slot >= this.items.length) {
			return false;
		}

		this.items[slot] = item;

		if (send) {
			this.onSlotChange(slot);
		}

		return true;
	}

	public Item getItem(int slot) {
		if (slot < 0 || slot >= this.items.length) {
			return Material.AIR.getItem();
		}

		return ItemUtils.getAirIfNull(this.items[slot]);
	}

	public void addItem(Item item) {
		for (int slot = 0; slot < this.items.length; slot++) {
			if (this.items[slot] == null) {
				this.setItem(slot, item);

				return;
			}
		}
	}

	public void clear() {
		this.clear(true);
	}

	public void clear(boolean send) {
		Arrays.fill(this.items, null);

		if (send) {
			this.sendSlots();
		}
	}

	public boolean openFor(Player player) {
		if (this.viewers.add(player)) {
			this.onOpen(player);

			return true;
		}

		return false;
	}

	protected void onOpen(Player player) {/**/}

	public boolean closeFor(Player player) {
		if (this.viewers.remove(player)) {
			this.onClose(player);

			return true;
		}

		return false;
	}

	protected void onClose(Player player) {
		var packet = new ContainerClosePacket();
		packet.setId((byte) this.getId());
		player.sendPacket(packet);
	}

	public void onSlotChange(int slot) {
		this.sendSlot(slot, this.viewers.toArray(new Player[0]));
	}

	protected void sendSlot(int slot, Player... players) {
		var packet = new InventorySlotPacket();
		packet.setContainerId(this.getId());
		packet.setSlot(slot);
		packet.setItem(ItemUtils.toNetwork(this.getItem(slot)));

		Arrays.stream(players).forEach(player -> player.sendPacket(packet));
	}

	public void sendSlots(Player player) {
		var packet = new InventoryContentPacket();
		packet.setContainerId(this.getId());
		packet.setContents(Arrays.stream(this.items)
				.map(item -> ItemUtils.toNetwork(ItemUtils.getAirIfNull(item)))
				.toList());
		player.sendPacket(packet);
	}

	public void sendSlots() {
		this.viewers.forEach(this::sendSlots);
	}

	public Set<Player> getViewers() {
		return new HashSet<>(this.viewers);
	}
}

File: src/main/java/org/distril/beengine/inventory/Inventory.kt
package org.distril.beengine.inventory

import com.nukkitx.protocol.bedrock.packet.ContainerClosePacket
import com.nukkitx.protocol.bedrock.packet.InventoryContentPacket
import com.nukkitx.protocol.bedrock.packet.InventorySlotPacket
import org.distril.beengine.material.Material
import org.distril.beengine.material.item.Item
import org.distril.beengine.player.Player
import org.distril.beengine.util.ItemUtils
import java.util.*
import java.util.concurrent.atomic.AtomicInteger

abstract class Inventory(
	open val holder: InventoryHolder,
	val type: InventoryType,
	overrideId: Int? = null
) {

	val viewers = mutableSetOf<Player>()

	private val items = arrayOfNulls<Item>(type.size)

	protected val id = overrideId ?: NEXT_ID.incrementAndGet()

	fun setItem(slot: Int, item: Item?) = this.setItem(slot, item, true)

	open fun setItem(slot: Int, item: Item?, send: Boolean): Boolean {
		if (slot < 0 || slot >= this.items.size) return false

		this.items[slot] = item

		if (send) this.onSlotChange(slot)

		return true
	}

	open fun getItem(slot: Int): Item {
		return if (slot < 0 || slot >= this.items.size) Material.AIR.getItem()
		else ItemUtils.getAirIfNull(this.items[slot])
	}

	fun addItem(item: Item) {
		this.items.indices.forEach {
			if (this.items[it] == null) {
				this.setItem(it, item)
				return
			}
		}
	}

	open fun clear() = this.clear(true)

	fun clear(send: Boolean) {
		Arrays.fill(this.items, null)

		if (send) this.sendSlots()
	}

	fun openFor(player: Player): Boolean {
		if (this.viewers.add(player)) {
			this.onOpen(player)
			return true
		}

		return false
	}

	protected open fun onOpen(player: Player) {
		/**/
	}

	fun closeFor(player: Player): Boolean {
		if (this.viewers.remove(player)) {
			this.onClose(player)
			return true
		}

		return false
	}

	protected fun onClose(player: Player) {
		val packet = ContainerClosePacket()
		packet.id = this.id.toByte()

		player.sendPacket(packet)
	}

	fun onSlotChange(slot: Int) = this.sendSlot(slot, *this.viewers.toTypedArray())

	protected open fun sendSlot(slot: Int, vararg players: Player) {
		val packet = InventorySlotPacket()
		packet.containerId = this.id
		packet.slot = slot
		packet.item = ItemUtils.toNetwork(this.getItem(slot))

		players.forEach { it.sendPacket(packet) }
	}

	open fun sendSlots(player: Player) {
		val packet = InventoryContentPacket()
		packet.containerId = this.id
		packet.contents = this.items.map { ItemUtils.toNetwork(ItemUtils.getAirIfNull(it)) }

		player.sendPacket(packet)
	}

	fun sendSlots() = this.viewers.forEach { this.sendSlots(it) }

	companion object {

		private val NEXT_ID = AtomicInteger(0)
	}
}
